File: avro/avro-annotation-processor/src/main/java/ru/tinkoff/kora/avro/annotation/processor/extension/AvroExtensionFactory.java
package ru.tinkoff.kora.avro.annotation.processor.extension;

import ru.tinkoff.kora.avro.annotation.processor.AvroTypes;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionFactory;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import java.util.Optional;

public class AvroExtensionFactory implements ExtensionFactory {

    @Override
    public Optional<KoraExtension> create(ProcessingEnvironment processingEnvironment) {
        var avro = processingEnvironment.getElementUtils().getTypeElement(AvroTypes.avroBinary.canonicalName());
        if (avro == null) {
            return Optional.empty();
        } else {
            return Optional.of(new AvroExtension(processingEnvironment));
        }
    }
}

File: avro/avro-symbol-processor/src/main/kotlin/ru/tinkoff/kora/avro/symbol/processor/extension/AvroExtensionFactory.kt
package ru.tinkoff.kora.avro.symbol.processor.extension

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import ru.tinkoff.kora.avro.symbol.processor.AvroTypes
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionFactory
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension

class AvroExtensionFactory : ExtensionFactory {

    override fun create(resolver: Resolver, kspLogger: KSPLogger, codeGenerator: CodeGenerator): KoraExtension? {
        val avro = resolver.getClassDeclarationByName(AvroTypes.avroBinary.canonicalName)
        return if (avro == null) {
            null
        } else {
            AvroKoraExtension(resolver, kspLogger, codeGenerator)
        }
    }
}
